src/v/kafka/server/handlers/produce.cc

```diff
@@ -190,6 +179,63 @@ static partition_produce_stages partition_append(
     };
 }
 
+static ss::future<model::record_batch> try_decompress(
+  bool is_compacted,
+  ss::foreign_ptr<std::unique_ptr<model::record_batch>> batch_ptr) {
+    auto batch = batch_ptr->copy();
+    if (is_compacted && batch.compressed()) {
+        return ss::do_with(std::move(batch), [](model::record_batch& batch) {
+            return storage::internal::decompress_batch(batch).then(
+              [&batch](model::record_batch&& b) {
+                  batch.decompressed(std::move(b));
+                  return std::move(batch);
+              });
+        });
+    }
+    return ss::make_ready_future<model::record_batch>(std::move(batch));
+}
+
+static std::optional<produce_response::partition>
+validate_keys(model::ntp& ntp, model::record_batch& batch) {
+    std::optional<produce_response::partition> error_opt;
+
+    std::vector<int32_t> null_keys;
+    if (batch.decompressed()) {
+        batch.decompressed()->for_each_record(
+          [&null_keys](const model::record& r) {
+              if (r.key_size() < 0) {
+                  null_keys.push_back(r.offset_delta());
+              }
+          });
+    } else if (!batch.compressed()) {
```

Comment:
can we work on the record_batch interface to make this a little more intuitive to read? For example, `if (batched.decompressed())` reads as if it is equivalent to the else condition `if (!batch.compressed())`